File: ios/FprKit/Sources/FprKit/CFBReader.swift
import Foundation

/// Reader for the Compound File Binary format ([MS-CFB]).
///
/// Both encrypted OOXML documents and Word/Excel 97-2003 files are CFB
/// containers: a little FAT filesystem inside a single file. Only the reading
/// half is implemented, because that is all decryption needs.
struct CFBReader {
    private static let signature: [UInt8] = [0xD0, 0xCF, 0x11, 0xE0, 0xA1, 0xB1, 0x1A, 0xE1]
    private static let endOfChain: UInt32 = 0xFFFF_FFFE
    private static let freeSector: UInt32 = 0xFFFF_FFFF

    struct DirectoryEntry {
        let name: String
        let type: UInt8       // 0 empty, 1 storage, 2 stream, 5 root
        let startSector: UInt32
        let size: UInt64
    }

    private let data: Data
    private let sectorSize: Int
    private let miniSectorSize: Int
    private let miniStreamCutoff: UInt32
    private let fat: [UInt32]
    private let miniFAT: [UInt32]
    private let directory: [DirectoryEntry]
    private let miniStream: Data

    static func isCFB(_ data: Data) -> Bool {
        data.count >= 8 && [UInt8](data.prefix(8)) == signature
    }

    init(_ data: Data) throws {
        guard Self.isCFB(data) else {
            throw FprError.corruptFile("not a compound file (bad signature)")
        }
        self.data = data

        var header = ByteReader(data, offset: 30)
        sectorSize = 1 << Int(try header.u16())
        miniSectorSize = 1 << Int(try header.u16())
        guard sectorSize >= 128, sectorSize <= 1 << 20 else {
            throw FprError.corruptFile("implausible sector size \(sectorSize)")
        }

        try header.seek(44)
        let fatSectorCount = Int(try header.u32())
        let firstDirectorySector = try header.u32()
        try header.seek(56)
        miniStreamCutoff = try header.u32()
        let firstMiniFATSector = try header.u32()
        let miniFATSectorCount = Int(try header.u32())
        let firstDIFATSector = try header.u32()
        let difatSectorCount = Int(try header.u32())

        // --- DIFAT: the list of sectors that hold the FAT.
        var difat: [UInt32] = []
        var entry = ByteReader(data, offset: 76)
        for _ in 0..<109 {
            let sector = try entry.u32()
            if sector == Self.freeSector { break }
            difat.append(sector)
        }
        var nextDIFAT = firstDIFATSector
        var difatGuard = 0
        while nextDIFAT != Self.endOfChain, nextDIFAT != Self.freeSector,
              difatGuard < difatSectorCount + 1 {
            var reader = ByteReader(data, offset: Self.offset(of: nextDIFAT, sectorSize))
            for _ in 0..<((sectorSize / 4) - 1) {
                let sector = try reader.u32()
                if sector != Self.freeSector { difat.append(sector) }
            }
            nextDIFAT = try reader.u32()
            difatGuard += 1
        }

        // --- FAT: the sector chains.
        var fatEntries: [UInt32] = []
        fatEntries.reserveCapacity(fatSectorCount * sectorSize / 4)
        for sector in difat.prefix(max(fatSectorCount, difat.count)) {
            var reader = ByteReader(data, offset: Self.offset(of: sector, sectorSize))
            for _ in 0..<(sectorSize / 4) {
                fatEntries.append((try? reader.u32()) ?? Self.freeSector)
            }
        }
        fat = fatEntries

        // --- MiniFAT, for streams below the cutoff.
        miniFAT = try Self.readChainValues(
            data: data, fat: fat, start: firstMiniFATSector,
            sectorSize: sectorSize, limit: miniFATSectorCount
        )

        // --- Directory.
        let directoryBytes = try Self.readChain(
            data: data, fat: fat, start: firstDirectorySector, sectorSize: sectorSize
        )
        var entries: [DirectoryEntry] = []
        var cursor = 0
        while cursor + 128 <= directoryBytes.count {
            let block = directoryBytes.subdata(
                in: (directoryBytes.startIndex + cursor)..<(directoryBytes.startIndex + cursor + 128)
            )
            var reader = ByteReader(block, offset: 64)
            let nameLength = Int(try reader.u16())
            let type = try reader.u8()
            try reader.seek(116)
            let startSector = try reader.u32()
            let size = try reader.u64()
            let rawName = block.prefix(max(0, min(64, nameLength - 2)))
            let name = String(data: Data(rawName), encoding: .utf16LittleEndian) ?? ""
            entries.append(
                DirectoryEntry(name: name, type: type, startSector: startSector, size: size)
            )
            cursor += 128
        }
        directory = entries

        // --- The mini stream itself lives in the root entry.
        if let root = entries.first(where: { $0.type == 5 }), root.size > 0 {
            miniStream = try Self.readChain(
                data: data, fat: fat, start: root.startSector, sectorSize: sectorSize
            ).prefix(Int(root.size))
        } else {
            miniStream = Data()
        }
    }

    /// Stream contents by name. Names are unique enough in the containers we
    /// care about that a flat search beats walking the red-black tree.
    func stream(named name: String) throws -> Data {
        guard let entry = directory.first(where: { $0.type == 2 && $0.name == name }) else {
            throw FprError.corruptFile("stream '\(name)' not found in compound file")
        }
        if entry.size < UInt64(miniStreamCutoff) {
            var output = Data()
            var sector = entry.startSector
            var guardCount = 0
            while sector != Self.endOfChain, sector != Self.freeSector,
                  guardCount < miniFAT.count + 1 {
                let start = Int(sector) * miniSectorSize
                guard start + miniSectorSize <= miniStream.count else { break }
                output.append(
                    miniStream.subdata(
                        in: (miniStream.startIndex + start)..<(miniStream.startIndex + start + miniSectorSize)
                    )
                )
                sector = Int(sector) < miniFAT.count ? miniFAT[Int(sector)] : Self.endOfChain
                guardCount += 1
            }
            return output.prefix(Int(entry.size))
        }
        return try Self.readChain(
            data: data, fat: fat, start: entry.startSector, sectorSize: sectorSize
        ).prefix(Int(entry.size))
    }

    func hasStream(named name: String) -> Bool {
        directory.contains { $0.type == 2 && $0.name == name }
    }

    private static func offset(of sector: UInt32, _ sectorSize: Int) -> Int {
        (Int(sector) + 1) * sectorSize
    }

    private static func readChain(
        data: Data, fat: [UInt32], start: UInt32, sectorSize: Int
    ) throws -> Data {
        var output = Data()
        var sector = start
        var visited = 0
        while sector != endOfChain, sector != freeSector, visited <= fat.count {
            let begin = offset(of: sector, sectorSize)
            guard begin + sectorSize <= data.count else { break }
            output.append(
                data.subdata(in: (data.startIndex + begin)..<(data.startIndex + begin + sectorSize))
            )
            guard Int(sector) < fat.count else { break }
            sector = fat[Int(sector)]
            visited += 1
        }
        return output
    }

    private static func readChainValues(
        data: Data, fat: [UInt32], start: UInt32, sectorSize: Int, limit: Int
    ) throws -> [UInt32] {
        let bytes = try readChain(data: data, fat: fat, start: start, sectorSize: sectorSize)
        var reader = ByteReader(bytes)
        var values: [UInt32] = []
        while reader.remaining >= 4, let value = try? reader.u32() { values.append(value) }
        return values
    }
}

File: android/fprkit/src/main/kotlin/dev/jayanth/fpr/CfbReader.kt
package dev.jayanth.fpr

import java.io.ByteArrayOutputStream

/**
 * Reader for the Compound File Binary format ([MS-CFB]).
 *
 * Encrypted OOXML documents and Word/Excel 97-2003 files are both CFB
 * containers: a small FAT filesystem inside one file. Only reading is
 * implemented, because that is all decryption needs.
 */
class CfbReader(private val data: ByteArray) {

    data class DirectoryEntry(
        val name: String,
        val type: Int,          // 0 empty, 1 storage, 2 stream, 5 root
        val startSector: Long,
        val size: Long,
    )

    private val sectorSize: Int
    private val miniSectorSize: Int
    private val miniStreamCutoff: Long
    private val fat: LongArray
    private val miniFat: LongArray
    private val directory: List<DirectoryEntry>
    private val miniStream: ByteArray

    init {
        require(isCfb(data)) { "not a compound file" }
        val header = ByteReader(data, 30)
        sectorSize = 1 shl header.u16()
        miniSectorSize = 1 shl header.u16()
        if (sectorSize < 128 || sectorSize > (1 shl 20)) {
            throw FprException.CorruptFile("implausible sector size $sectorSize")
        }

        header.seek(44)
        val fatSectorCount = header.u32().toInt()
        val firstDirectorySector = header.u32()
        header.seek(56)
        miniStreamCutoff = header.u32()
        val firstMiniFatSector = header.u32()
        val miniFatSectorCount = header.u32().toInt()
        val firstDifatSector = header.u32()
        val difatSectorCount = header.u32().toInt()

        val difat = ArrayList<Long>()
        val entry = ByteReader(data, 76)
        repeat(109) {
            val sector = entry.u32()
            if (sector != FREE_SECTOR) difat.add(sector)
        }
        var nextDifat = firstDifatSector
        var guard = 0
        while (nextDifat != END_OF_CHAIN && nextDifat != FREE_SECTOR && guard <= difatSectorCount) {
            val reader = ByteReader(data, sectorOffset(nextDifat))
            repeat(sectorSize / 4 - 1) {
                val sector = reader.u32()
                if (sector != FREE_SECTOR) difat.add(sector)
            }
            nextDifat = reader.u32()
            guard += 1
        }

        val fatValues = ArrayList<Long>(fatSectorCount * sectorSize / 4)
        for (sector in difat) {
            val reader = ByteReader(data, sectorOffset(sector))
            repeat(sectorSize / 4) {
                fatValues.add(runCatching { reader.u32() }.getOrDefault(FREE_SECTOR))
            }
        }
        fat = fatValues.toLongArray()

        miniFat = readChain(firstMiniFatSector).let { bytes ->
            val reader = ByteReader(bytes)
            val values = ArrayList<Long>()
            while (reader.remaining >= 4) values.add(reader.u32())
            values.toLongArray()
        }

        val directoryBytes = readChain(firstDirectorySector)
        val entries = ArrayList<DirectoryEntry>()
        var cursor = 0
        while (cursor + 128 <= directoryBytes.size) {
            val block = directoryBytes.copyOfRange(cursor, cursor + 128)
            val reader = ByteReader(block, 64)
            val nameLength = reader.u16()
            val type = reader.u8()
            reader.seek(116)
            val startSector = reader.u32()
            val size = reader.u64()
            val nameBytes = block.copyOfRange(0, maxOf(0, minOf(64, nameLength - 2)))
            entries.add(
                DirectoryEntry(String(nameBytes, Charsets.UTF_16LE), type, startSector, size)
            )
            cursor += 128
        }
        directory = entries

        val root = entries.firstOrNull { it.type == 5 }
        miniStream = if (root != null && root.size > 0) {
            readChain(root.startSector).copyOf(minOf(root.size.toInt(), Int.MAX_VALUE))
        } else {
            ByteArray(0)
        }
    }

    fun hasStream(name: String): Boolean = directory.any { it.type == 2 && it.name == name }

    fun stream(name: String): ByteArray {
        val entry = directory.firstOrNull { it.type == 2 && it.name == name }
            ?: throw FprException.CorruptFile("stream '$name' not found in compound file")
        if (entry.size < miniStreamCutoff) {
            val output = ByteArrayOutputStream()
            var sector = entry.startSector
            var guard = 0
            while (sector != END_OF_CHAIN && sector != FREE_SECTOR && guard <= miniFat.size) {
                val start = sector.toInt() * miniSectorSize
                if (start + miniSectorSize > miniStream.size) break
                output.write(miniStream, start, miniSectorSize)
                sector = if (sector < miniFat.size) miniFat[sector.toInt()] else END_OF_CHAIN
                guard += 1
            }
            return output.toByteArray().copyOf(entry.size.toInt())
        }
        return readChain(entry.startSector).copyOf(entry.size.toInt())
    }

    private fun sectorOffset(sector: Long): Int = ((sector + 1) * sectorSize).toInt()

    private fun readChain(start: Long): ByteArray {
        val output = ByteArrayOutputStream()
        var sector = start
        var visited = 0
        while (sector != END_OF_CHAIN && sector != FREE_SECTOR && visited <= fat.size) {
            val begin = sectorOffset(sector)
            if (begin + sectorSize > data.size) break
            output.write(data, begin, sectorSize)
            if (sector >= fat.size) break
            sector = fat[sector.toInt()]
            visited += 1
        }
        return output.toByteArray()
    }

    companion object {
        private const val END_OF_CHAIN = 0xFFFFFFFEL
        private const val FREE_SECTOR = 0xFFFFFFFFL
        private val SIGNATURE = byteArrayOf(
            0xD0.toByte(), 0xCF.toByte(), 0x11, 0xE0.toByte(),
            0xA1.toByte(), 0xB1.toByte(), 0x1A, 0xE1.toByte(),
        )

        fun isCfb(data: ByteArray): Boolean =
            data.size >= 8 && SIGNATURE.indices.all { data[it] == SIGNATURE[it] }
    }
}
